### PR TITLE
8213905: reflection not working for type annotations applied to exception types in the inner class constructor

### PR DESCRIPTION
**Pull request by vicente-romero-oracle**
Hi,

Please review this change that is fixing a bug in reflection in particular in `sun.reflect.annotation.TypeAnnotationParser::buildAnnotatedTypes` the current code is assuming that for inner class constructors, there are only type annotations on parameters, but it is also invoked to extract type annotations applied to exception types for example. This bug affects the behavior of method: `java.lang.reflect.Executable::getAnnotatedExceptionTypes` which is not behaving according to its specification. Given that this fix affects the behavior of a method belonging to our API a CSR has been filed too. Please review it at [JDK-8278926](https://bugs.openjdk.java.net/browse/JDK-8278926).

TIA

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed
- [x] Change requires a CSR request to be approved

### Issues
 * [JDK-8213905](https://bugs.openjdk.java.net/browse/JDK-8213905): reflection not working for type annotations applied to exception types in the inner class constructor
 * [JDK-8278926](https://bugs.openjdk.java.net/browse/JDK-8278926): reflection not working for type annotations applied to exception types in the inner class constructor (**CSR**)

### Reviewers
 * [Jan Lahoda](https://openjdk.java.net/census#jlahoda) (@lahodaj - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/6869/head:pull/6869` \
`$ git checkout pull/6869`

Update a local copy of the PR: \
`$ git checkout pull/6869` \
`$ git pull https://git.openjdk.java.net/jdk pull/6869/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 6869`

View PR using the GUI difftool: \
`$ git pr show -t 6869`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/6869.diff">https://git.openjdk.java.net/jdk/pull/6869.diff</a>

</details>

### CODE REVIEW

**Comment by jddarcy** [comment]
Hi Vicente.

Please file a CSR for the behavioral change.

**Comment by vicente-romero-oracle** [comment]
> Hi Vicente.
> 
> Please file a CSR for the behavioral change.

sure, thanks

File: src/java.base/share/classes/sun/reflect/annotation/TypeAnnotationParser.java [hunk 1]
```diff
@@ -132,7 +132,8 @@ public static AnnotatedType buildAnnotatedType(byte[] rawAnnotations,
             Class<?> declaringClass = ctor.getDeclaringClass();
             if (!declaringClass.isEnum() &&
                 (declaringClass.isMemberClass() &&
-                 (declaringClass.getModifiers() & Modifier.STATIC) == 0) ) {
+                 (declaringClass.getModifiers() & Modifier.STATIC) == 0) &&
+                 filter == TypeAnnotation.TypeAnnotationTarget.METHOD_FORMAL_PARAMETER) {
                 offset = true;
```
Comment (liach) [inline src/java.base/share/classes/sun/reflect/annotation/TypeAnnotationParser.java:137]: Currently, we have a few parallel systems for determining if certain parameters are synthetic. This guessing by being inner classes or enums is one, and detecting the flags from `MethodParameters` attribute (used for handling signatures) is another. (In fact, both are present in `Executable` in `handleParameterNumberMismatch` and `getAllGenericParameterTypes`)

Detecting `MethodParameters` attribute should be more reliable, especially for other code that may include methods with any number of synthetic or mandated parameters (presumably in the beginning of the parameter list). However, javac will not produce that attribute without `-parameters` flag, so most class files compiled to be missing this attribute.

IMO a longer-term solution is to opt-in Javac to emit critical method parameter information by default (ex. emit the parameter flags without names) whenever it emits parameter signatures or annotations. And then the system for skipping synthetic/mandated parameters in the two places may possibly be unified.

**Review by lahodaj — APPROVED** [review]
To me, looks good.

**Comment by vicente-romero-oracle** [comment]
> To me, looks good.

thanks

**Pull request closed without merging**